quickfixj-core/src/main/java/quickfix/JdbcUtil.java

```diff
@@ -31,89 +33,96 @@
+        String jdbcDriver = settings.getString(sessionID, JdbcSetting.SETTING_JDBC_DRIVER);
+        String connectionURL = settings.getString(sessionID,JdbcSetting.SETTING_JDBC_CONNECTION_URL);
+        String user = settings.getString(sessionID, JdbcSetting.SETTING_JDBC_USER);
+        String password = settings.getString(sessionID, JdbcSetting.SETTING_JDBC_PASSWORD);
+        return getOrCreatePooledDataSource(settings, sessionID, jdbcDriver, connectionURL, user, password);
     }
 
-    /**
-     * This is typically called from a single thread, but just in case we are using an atomic loading function
-     * to avoid the creation of two data sources simultaneously. The cache itself is thread safe.
-     */
-    static DataSource getDataSource(String jdbcDriver, String connectionURL, String user, String password,
-                                                 boolean cache, int maxConnCount, int simultaneousBuildThrottle,
-                                                 long maxActiveTime, int maxConnLifetime) {
+    static DataSource getOrCreatePooledDataSource(SessionSettings settings, SessionID sessionID, String jdbcDriver, String connectionURL, String user, String password)
+            throws ConfigError, FieldConvertError {
         String key = jdbcDriver + "#" + connectionURL + "#" + user + "#" + password;
-        ProxoolDataSource ds = cache ? dataSources.get(key) : null;
-
-        if (ds == null) {
-            final Function<String, ProxoolDataSource> loadingFunction = dataSourceKey -> {
-                final ProxoolDataSource dataSource = new ProxoolDataSource(CONNECTION_POOL_ALIAS + "-" + dataSourceCounter.incrementAndGet());
-
-                dataSource.setDriver(jdbcDriver);
-                dataSource.setDriverUrl(connectionURL);
-
-                // Bug in Proxool 0.9RC2. Must set both delegate properties and individual setters. :-(
-                dataSource.setDelegateProperties("user=" + user + ","
-                        + (password != null && !"".equals(password) ? "password=" + password : ""));
-                dataSource.setUser(user);
-                dataSource.setPassword(password);
-
-                dataSource.setMaximumActiveTime(maxActiveTime);
-                dataSource.setMaximumConnectionLifetime(maxConnLifetime);
-                dataSource.setMaximumConnectionCount(maxConnCount);
-                dataSource.setSimultaneousBuildThrottle(simultaneousBuildThrottle);
-                return dataSource;
-            };
-            ds = cache ? dataSources.computeIfAbsent(key, loadingFunction) : loadingFunction.apply(key);
+
+        HikariDataSource dataSource = dataSources.get(key);
+
+        if (dataSource != null) {
+            return dataSource;
+        }
+
+        HikariDataSource newDataSource = createPooledDataSource(settings, sessionID, jdbcDriver, connectionURL, user, password);
+
+        if (dataSources.putIfAbsent(key, newDataSource) == null) {
```

Comment:
Map loading function was replaced with raw code as SessionSettings throws checked exceptions and we probably want to access them when DataSource is actually created.